modal/parallel_map.py

```diff
@@ -450,36 +473,68 @@ async def _map_async(
     prefer: throughput (prioritize queueing inputs) or latency (prioritize yielding results)
     """
+
     raw_input_queue: Any = SynchronizedQueue()  # type: ignore
     raw_input_queue.init()
-
-    async def feed_queue():
-        # This runs in a main thread event loop, so it doesn't block the synchronizer loop
-        async with aclosing(async_zip(*[sync_or_async_iter(it) for it in input_iterators])) as streamer:
-            async for args in streamer:
-                await raw_input_queue.put.aio((args, kwargs))
-        await raw_input_queue.put.aio(None)  # end-of-input sentinel
-        if False:
-            # make this a never yielding generator so we can async_merge it below
-            # this is important so any exception raised in feed_queue will be propagated
-            yield
+    feed_queue_gen = _feed_raw_input_queue_from_iterators(raw_input_queue, *input_iterators, kwargs=kwargs)
 
     # note that `map()` and `map.aio()` are not synchronicity-wrapped, since
     # they accept executable code in the form of
     # iterators that we don't want to run inside the synchronicity thread.
     # Instead, we delegate to `._map()` with a safer Queue as input
     async with aclosing(
-        async_merge(self._map.aio(raw_input_queue, order_outputs, return_exceptions), feed_queue())
+        async_merge(self._map.aio(raw_input_queue, order_outputs, return_exceptions), feed_queue_gen)
     ) as map_output_stream:
         async for output in map_output_stream:
             yield output
 
 
+async def _spawn_map_generator(self, *input_iterators, kwargs={}) -> typing.AsyncGenerator[None, None]:
+    """mdmd:hidden
+    This runs in an event loop on the main thread. It consumes inputs
+    from the input iterators and creates async function calls for each.
+
+    We create a common generator that is shared by `_spawn_map_async` and `_spawn_map_sync` because
+    it is easier to wrap async generators than coroutines.
+    """
+
+    raw_input_queue: Any = SynchronizedQueue()  # type: ignore
+    raw_input_queue.init()
+    feed_queue_gen = _feed_raw_input_queue_from_iterators(raw_input_queue, *input_iterators, kwargs=kwargs)
+
+    async with aclosing(
+        async_merge(
+            self._spawn_map_inner.aio(raw_input_queue),
+            feed_queue_gen,
+        )
+    ) as streamer:
+        async for res in streamer:
+            yield res
+            return
+
+
+async def _spawn_map_async(self, *input_iterators, kwargs={}) -> None:
+    async for res in _spawn_map_generator(self, *input_iterators, kwargs=kwargs):
+        return res
+
+
+def _spawn_map_sync(self, *input_iterators, kwargs={}) -> None:
+    return list(
+        AsyncOrSyncIterable(
+            _spawn_map_generator(self, *input_iterators, kwargs=kwargs),
+            nested_async_message=(
+                "You can't run Function.spawn_map() from an async function. Use Function.map.aio() instead."
```

Comment:
```suggestion
                "You can't run Function.spawn_map() from an async function. Use Function.spawn_map.aio() instead."
```

Reply:
Why is this the only place we disarm this footgun?